pdns/dnsdistdist/doh.cc

```diff
@@ -0,0 +1,718 @@
+#include "dnsdist.hh"
+#include "misc.hh"
+#include <thread>
+#include "dns.hh"
+#include "dolog.hh"
+#include "dnsdist-ecs.hh"
+#include "dnsdist-rules.hh"
+#include "dnsdist-xpf.hh"
+#include <boost/algorithm/string.hpp>
+
+using namespace std;
+
+/* So, how does this work. We use h2o for our http2 and TLS needs. 
+   If the operator has configured multiple IP addresses to listen on, 
+   we launch multiple h2o listener threads. We can hook in to multiple 
+   URLs though on the same IP. There is no SNI yet (I think).
+
+   h2o is event driven, so we get callbacks if a new DNS query arrived.
+   When it does, we do some minimal parsing on it, and send it on to the 
+   dnsdist worker thread which we also launched.
+
+   This dnsdist worker thread injects the query into the normal dnsdist flow 
+   (as a datagram over a socketpair). The response also goes back over a 
+   (different) socketpair, where we pick it up and deliver it back to h2o.
+
+   For coordination, we use the h2o socket multiplexer, which is sensitive to our
+   socketpair too.
+*/
+
+/* h2o notes.
+   Paths and parameters etc just *happen* to be null-terminated in HTTP2.
+   They are not in HTTP1. So you MUST use the length field! 
+*/
+
+// we create on of these per thread, and pass around a pointer to it
+// through the bowels of h2o
+struct DOHServerConfig
+{
+  h2o_globalconf_t h2o_config;
+  h2o_context_t h2o_ctx;
+  h2o_accept_ctx_t h2o_accept_ctx;
+
+  int dohquerypair[2];
+  int dohresponsepair[2];
+  ClientState cs;
+  std::shared_ptr<DOHFrontend> df;
+};
+
+/* this duplicates way too much from the UDP handler. Sorry.
+   this function calls 'return -1' to drop a query without sending it
+   caller should make sure HTTPS thread hears of that
+*/
+
+static int processDOHQuery(DOHUnit* du)
+{
+  LocalHolders holders;
+  uint16_t queryId=0;
+  try {
+    if(!du->req) // we got closed meanwhile. XXX small race condition here
+      return -1;
+    DOHServerConfig* dsc = (DOHServerConfig*)du->req->conn->ctx->storage.entries[0].data;
+    ClientState& cs = dsc->cs;
+    
+    /* we need an accurate ("real") value for the response and
+       to store into the IDS, but not for insertion into the
+       rings for example */
+    struct timespec queryRealTime;
+    struct timespec now;
+    gettime(&now);
+    gettime(&queryRealTime, true);
+    char* query = (char*) du->query.c_str();
+
+    struct dnsheader* dh = reinterpret_cast<struct dnsheader*>((char*)query);
+    uint16_t len = du->query.length();
+
+    if (!checkQueryHeaders(dh)) {
+      return -1; // drop
+    }
+
+    string poolname;
+    int delayMsec = 0;
+    const uint16_t * flags = getFlagsFromDNSHeader(dh);
+    const uint16_t origFlags = *flags;
+    uint16_t qtype, qclass;
+    unsigned int consumed = 0;
+    DNSName qname(query, len, sizeof(dnsheader), false, &qtype, &qclass, &consumed);
+    DNSQuestion dq(&qname, qtype, qclass, consumed, &du->dest, &du->remote, dh, 1500, len, false, &queryRealTime);
```

Comment:
The `1500` here is wrong, it pretends that there is at least 1500 bytes available from the address pointed to by `dh`, which is not true. Rules and `handleEDNSClientSubnet()` will very likely trigger an heap-based out-of-bounds write.